src/frontend/src/optimizer/plan_node/logical_project_set.rs

```diff
@@ -350,3 +372,58 @@ impl ToStream for LogicalProjectSet {
         Ok(StreamProjectSet::new(new_logical).into())
     }
 }
+
+#[cfg(test)]
+mod test {
+    use risingwave_common::catalog::{Field, Schema};
+    use risingwave_common::types::DataType;
+
+    use super::*;
+    use crate::expr::{ExprImpl, InputRef, TableFunction};
+    use crate::optimizer::plan_node::LogicalValues;
+    use crate::optimizer::property::FunctionalDependency;
+    use crate::session::OptimizerContext;
+
+    #[tokio::test]
+    async fn fd_derivation_project_set() {
+        let ctx = OptimizerContext::mock().await;
+        let fields: Vec<Field> = vec![
+            Field::with_name(DataType::Int32, "v1"),
+            Field::with_name(DataType::Int32, "v2"),
+            Field::with_name(DataType::Int32, "v3"),
+        ];
+        let mut values = LogicalValues::new(vec![], Schema { fields }, ctx);
+        values
+            .base
+            .functional_dependency
+            .add_functional_dependency_by_column_indices(&[1], &[2], 3);
+        let project_set = LogicalProjectSet::new(
+            values.into(),
+            vec![
+                ExprImpl::InputRef(Box::new(InputRef::new(2, DataType::Int32))),
+                ExprImpl::InputRef(Box::new(InputRef::new(1, DataType::Int32))),
+                ExprImpl::TableFunction(Box::new(
+                    TableFunction::new(
+                        crate::expr::TableFunctionType::Generate,
+                        vec![
+                            ExprImpl::InputRef(Box::new(InputRef::new(0, DataType::Int32))),
+                            ExprImpl::InputRef(Box::new(InputRef::new(1, DataType::Int32))),
+                            ExprImpl::InputRef(Box::new(InputRef::new(2, DataType::Int32))),
+                        ],
+                    )
+                    .unwrap(),
+                )),
+            ],
+        );
+        let fd_set = project_set.base.functional_dependency;
+        let expected_fd_set = vec![FunctionalDependency::with_indices(4, &[0, 2], &[1])];
+        assert_eq!(fd_set.as_dependencies().len(), expected_fd_set.len());
+        for i in fd_set.as_dependencies() {
+            assert!(
+                expected_fd_set.contains(i),
+                "{} should be in expected_fd_set",
+                i
+            );
+        }
```

Comment:
What about assert_eq 2 Vecs?

Reply:
Btw this iteration using `contains` might be just be unexpected. e.g., `actual=vec![1,1,1]` & `expected=vec![1,2,3]`

Reply:
Also for other nodes

Reply:
> What about assert_eq 2 Vecs?

I'm worried about the order of the dependencies. Maybe we can use a HashSet.

Reply:
We can use either sorting or hashset.

Reply:
Hashset might be slightly better